code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -395,3 +394,196 @@
 	component_parts += new /obj/item/stack/sheet/glass(null)
 	component_parts += new cell_type(null)
 	RefreshParts()
+
+// Handheld chem dispenser
+/obj/item/handheld_chem_dispenser
+	name = "handheld chem dispenser"
+	icon = 'icons/obj/chemical.dmi'
+	item_state = "handheld_chem"
+	icon_state = "handheld_chem"
+	flags = NOBLUDGEON
+	var/ui_title = "Handheld Chem Dispenser"
+	var/obj/item/stock_parts/cell/high/cell = null
+	var/amount = 10
+	var/mode = "dispense"
+	var/is_drink = FALSE
+	var/list/dispensable_reagents = list("hydrogen", "lithium", "carbon", "nitrogen", "oxygen", "fluorine",
+	"sodium", "aluminum", "silicon", "phosphorus", "sulfur", "chlorine", "potassium", "iron",
+	"copper", "mercury", "plasma", "radium", "water", "ethanol", "sugar", "iodine", "bromine", "silver", "chromium")
+	var/current_reagent = null
+	var/efficiency = 0.2
+	var/recharge_rate = 1 // Keep this as an integer
+
+/obj/item/handheld_chem_dispenser/New()
+	..()
+	cell = new(src)
+	dispensable_reagents = sortList(dispensable_reagents)
+	current_reagent = pick(dispensable_reagents)
+	update_icon()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/handheld_chem_dispenser/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+/obj/item/handheld_chem_dispenser/get_cell()
+	return cell
+
+/obj/item/handheld_chem_dispenser/afterattack(obj/target, mob/user, proximity)
+	if(!proximity)
+		return
+
+	if(!current_reagent)
+		return
+
+	if(!amount)
+		return
+
+	if(!check_allowed_items(target,target_self = TRUE))
+		return
+
+	if(!target.is_refillable())
+		return
+
+	if(mode == "dispense")
+		var/free = target.reagents.maximum_volume - target.reagents.total_volume
+		var/actual = min(amount, cell.charge / efficiency, free)
+		target.reagents.add_reagent(current_reagent, actual)
+		cell.charge -= actual / efficiency
+		if(actual)
+			to_chat(usr, "You dispense [amount] units of [current_reagent] into the [target].")
+		update_icon()
+	else if(mode == "remove")
+		if(target.reagents.remove_reagent(current_reagent, amount))
+			to_chat(usr, "You remove [amount] units of [current_reagent] from the [target].")
+	else if(mode == "isolate")
+		if(target.reagents.isolate_reagent(current_reagent))
+			to_chat(usr, "You remove all but [current_reagent] from the [target].")
+
+/obj/item/handheld_chem_dispenser/attack_self(mob/user)
+	tgui_interact(user)
+
+
+/obj/item/handheld_chem_dispenser/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_inventory_state)
+	// update the ui if it exists, returns null if no ui is passed/found
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "HandheldChemDispenser", ui_title, 390, 500)
+		ui.open()
+
+/obj/item/handheld_chem_dispenser/tgui_data(mob/user)
+	var/data[0]
```

Comment:
```suggestion
	var/list/data = list()
```

Reply:
I copied how the TGUI for the chem dispenser was done as an example. I assumed that one was accurate. Changes will be made.